File: hacker-rank-interview-preparation-kit/Ruby/warm-up-challenges/repeated_string.rb
=begin
Tae H0 K1m
  4. Repeated String

  Lilah has a string, class << self; self end, of lowercase English letters that
  she repeated infinitely many times.

  Given an integer, n, find and print the number of letter a's in the first
  letters of Lilah's infinite string.
=end

# 1. count how many "a" in the string
# 2. find how many time s repeated
# 3. count rest of "a"s

def repeatedString(s, n)
  count = 0

  # 1) count how many "a" in the string
  for i in (0...s.length)
    count += 1 if s[i] == "a"
  end

  # 2) find how many time s repeated
  times = n / s.length
  count *= times

  # 3) count rest of "a"s
  for i in 0...(n % s.length)
    count += 1 if s[i] == "a"
  end

  return count
end

puts repeatedString("abcac", 10)
# => 4

puts repeatedString("aba", 10)
# => 7, abc abc abc a

puts repeatedString("a", 1_000_000_000_000)
# => 1_000_000_000_000

File: hacker-rank-interview-preparation-kit/Python/warm-up-challenges/repeated_string.py
""" #4 Repeated String
    Tae H0 K1m

    problem:

    input: a string which is going to repeated (consists of lowercase letters)
    output: an integer representing the number of ouccurrences of "a" in the
    given string

    arrproach:
        There is two part.

        1) Find how many "a" appears in the string
        2) Find how many "a" appears in the remaning string

"""

def repeatedString(s, n):
    num = 0

    # 1
    freq = {}
    for i in s:
        if i in freq:
            freq[i] += 1
        else:
            freq[i] = 1

    # check if the given string has the character "a"
    if "a" in freq:
        num += freq["a"] * int((n / len(s)))

        # 2
        for i in range(int((n % len(s)))):
            if s[i] == "a":
                num += 1

        return num
    else:
        return 0

print(repeatedString("abcac", 10))
# => 4

print(repeatedString("aba", 10))
# => 7

print(repeatedString("a", 1000000000000))
# => 1000000000000

print(repeatedString("ceebbcb", 817723))
# => 0?
